portal/common/lib/rpc_selector.ts

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+    GetDynamicFieldObjectParams,
+    GetObjectParams,
+    MultiGetObjectsParams,
+    SuiClient,
+    SuiObjectResponse,
+} from "@mysten/sui/client";
+import { TESTNET_RPC_LIST } from "./constants";
+
+interface RPCSelectorInterface {
+    getObject(input: GetObjectParams): Promise<SuiObjectResponse>;
+    multiGetObjects(input: MultiGetObjectsParams): Promise<SuiObjectResponse[]>;
+    getDynamicFieldObject(
+        input: GetDynamicFieldObjectParams,
+    ): Promise<SuiObjectResponse>;
+    call<T>(method: string, args: any[]): Promise<T>;
+}
+
+class RPCSelector implements RPCSelectorInterface {
+    private static instance: RPCSelector;
+    private clients: SuiClient[];
+    private selectedClient: SuiClient | undefined;
+
+    private constructor(rpcURLs: string[]) {
+        // Initialize clients.
+        this.clients = rpcURLs.map((rpcUrl) => new SuiClient({ url: rpcUrl }));
+        this.selectedClient = undefined;
+    }
+
+    // Get the singleton instance.
+    public static getInstance(): RPCSelector {
+    if (!RPCSelector.instance) {
+        RPCSelector.instance = new RPCSelector(TESTNET_RPC_LIST);
+    }
+    return RPCSelector.instance;
+    }
+
+    // General method to call clients and return the first successful response.
+    private async invokeWithFailover<T>(methodName: string, args: any[]): Promise<T> {
+        if (this.clients.length === 0) {
+            throw new Error("No available clients to handle the request.");
+        }
+
+        const isNoSelectedClient = !this.selectedClient;
+        if (isNoSelectedClient) {
+            return await this.callFallbackClients<T>(methodName, args);
+        }
+
+        try {
+            return await this.callSelectedClient<T>(methodName, args);
+        } catch (error) {
+            console.error(`Selected client failed: ${error}`);
+            this.selectedClient = undefined;
+            return await this.callFallbackClients<T>(methodName, args);
+        }
+    }
+
+    // Attempt to call the method on the selected client with a timeout.
+    private async callSelectedClient<T>(methodName: string, args: any[]): Promise<T> {
+        const method = (this.selectedClient as any)[methodName] as Function;
+        if (!method) {
+            throw new Error(`Method ${methodName} not found on selected client`);
+        }
+        const timeoutDuration = 5000;
```

Comment:
Remark: this is a bit arbitrary and may be too low, especially if the call needs to load a resource. 

Question: Is there a way to set a timeout on connection establishment, instead of on the whole call?

Reply:
It might be a bit challenging since the actual call happens through the Sui client, but I can extend the time.